File: source/go/file_defer_close.go
package main

import (
	"fmt"
	"os"
)

func main() {
	path := "/tmp/hemlock_defer_test.txt"
	err := os.WriteFile(path, []byte("hello from defer"), 0644)
	if err != nil {
		fmt.Println("write error:", err)
		return
	}

	f, err := os.Open(path)
	if err != nil {
		fmt.Println("open error:", err)
		return
	}
	defer f.Close()

	contents, err := os.ReadFile(path)
	if err != nil {
		fmt.Println("read error:", err)
		return
	}

	if string(contents) == "hello from defer" {
		fmt.Println("wrote and read back successfully")
	}
}

File: source/python/file_defer_close.py
import os

def main():
    path = "/tmp/hemlock_defer_test.txt"
    with open(path, "w") as f:
        f.write("hello from defer")

    with open(path, "r") as f:
        contents = f.read()
        if contents == "hello from defer":
            print("wrote and read back successfully")

main()
